lint-rules/src/main/java/com/duckduckgo/lint/strings/PlaceholderDetector.kt

```diff
@@ -0,0 +1,101 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.duckduckgo.lint.strings
+
+import com.android.SdkConstants
+import com.android.ide.common.resources.usage.ResourceUsageModel
+import com.android.tools.lint.detector.api.Category.Companion.CUSTOM_LINT_CHECKS
+import com.android.tools.lint.detector.api.Context
+import com.android.tools.lint.detector.api.Implementation
+import com.android.tools.lint.detector.api.Issue
+import com.android.tools.lint.detector.api.ResourceXmlDetector
+import com.android.tools.lint.detector.api.Scope
+import com.android.tools.lint.detector.api.Severity
+import com.android.tools.lint.detector.api.XmlContext
+import com.android.tools.lint.detector.api.XmlScanner
+import org.w3c.dom.Attr
+import org.w3c.dom.Document
+import org.w3c.dom.Node
+
+@Suppress("UnstableApiUsage")
+class PlaceholderDetector : ResourceXmlDetector(), XmlScanner {
+
+    private fun ignoreFile(context: Context) =
+        context.file.name.startsWith("donottranslate") ||
+            ResourceUsageModel.isAnalyticsFile(context.file) ||
+            !context.project.reportIssues
+
+
+    override fun visitDocument(context: XmlContext, document: Document) {
+        if (ignoreFile(context)) {
+            return
+        }
+
+        val root = document.documentElement
+        val translatable: Attr? = root?.getAttributeNode(SdkConstants.ATTR_TRANSLATABLE)
+        if (translatable?.value != SdkConstants.VALUE_FALSE) {
+            if (context.file.parentFile?.name == SdkConstants.FD_RES_VALUES) {
+                visitNode(context, document)
+            }
+            return
+        }
+    }
+
+    private fun visitNode(context: XmlContext, node: Node) {
+        val nodeType = node.nodeType
+        if (nodeType == Node.TEXT_NODE) {
+            val text = node.nodeValue
+            if (PLACEHOLDERS.firstOrNull { text.contains(it) } != null) {
+                context.report(
+                    PLACEHOLDER_MISSING_POSITION, node, location = context.getNameLocation(node),
+                    "Placeholder is missing the position in a string"
+                )
+            }
+        }
+
+        // Visit children
+        val childNodes = node.childNodes
+        var i = 0
+        val n = childNodes.length
+        while (i < n) {
+            val child = childNodes.item(i)
+            visitNode(context, child)
+            i++
+        }
+    }
+
+
+    companion object {
+        val PLACEHOLDERS = listOf(
+            "%s", "%d"
```

Comment:
nit: a niche case that might get missed: floats (which i think are valid placeholders)

Reply:
Oki so, this is interesting. For the record this is wrong, as you say there are many more placeholders and different scenarios but the logic is mind blowing to get it right so instead and to save time I added the typical ones that we use. I've also now added f and o. Should have probably added b to honest but follow up maybe.